### PR TITLE
Use QFormLayout for layer control grid

### PR DESCRIPTION
**Pull request by brisvag**
# Description
Lately I've been touching the qt layer controls several times in different PRs. Things often got finnicky because of this whole grid system with rows, columns, row spans and column spans. Indeed, I encountered a few cases where things were "broken" and just working by chance, but ready to fall apart at the first touch.

To solve this issue, I came up with a simplified interface. Might be overengineered, but I find it much more intuitive, less error prone and more readable.

`QtLayerControls` now has a `_populate_grid` method that takes a list of tuples containing strings, `QWidget`s, `...`s and `None`s, and constructs a grid accordingly. `Ellipsis` is intepreted as "increase the span of the widget before, or the row above". `None` is an empty space, and strings are simply made into (translated) labels. "Unfinished" row get right-padded with ellipses.

For example, this:
```
('a label', QSlider, ...),
(...,),
(None, 'another label'),
(QWidget, ..., None),
```

Will become this:
```
a label        | QSlider with column span 2         |
with row span 2| and row span 2                     |
-----------------------------------------------------
EMPTY          | another label with column span 2   |
-----------------------------------------------------
Qwidget with column span 2       | EMPTY            |
```

It's obviously not covering all edge cases, but it's probably good enough for this job. The main advantage of this approach is that it's easy to shuffle around things without having to change all the numbers manually.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by tlambert03** [comment]
hmm... I feel your pain.  I've definitely been annoyed by having to change those as well.   need to ponder this one a bit.  while I definitely agree that calling `_populate_grid` is easier to read, the code in _populate grid is definitely harder to parse than the current list of `addWidget` (and as such, probably needs some tests)

**Comment by brisvag** [comment]
Yeah, I agree. I just figured I'd do it for a couple of layers to see how it looks like and get some feedback here :)

EDIT: well, it was easy to convert everything once I did it once...

File: napari/_qt/layer_controls/qt_layer_controls_base.py
Comment (Carreau) [inline napari/_qt/layer_controls/qt_layer_controls_base.py]: I would make sure that here wdg is a widget, we can imagine someone passing `(..., something_else)`, which would be invalid.

File: napari/_qt/layer_controls/qt_layer_controls_base.py
Comment (Carreau) [inline napari/_qt/layer_controls/qt_layer_controls_base.py]: Same here, I would check that if it's neither it fails.

**Review by Carreau — APPROVED** [review]
unlike @tlambert03 I'm less worried about the difficulty of reading `_populate_grid`, though I'd like `_populate_grid` (or a utility function) to return something (args after line 138?) so that it's easier to test.

File: napari/_qt/layer_controls/qt_layer_controls_base.py
Comment (brisvag) [reply napari/_qt/layer_controls/qt_layer_controls_base.py]: Added a check.

File: napari/_qt/layer_controls/qt_layer_controls_base.py
Comment (brisvag) [reply napari/_qt/layer_controls/qt_layer_controls_base.py]: I think this is unnecessary once I added the above, since every element that gets here was put there by the previous loop.

**Comment by brisvag** [comment]
I addressed the comments; I also changed some variable names and used a `namedtuple` in order to make it more readable. Is it clearer @tlambert03 ?

**Comment by Carreau** [comment]
There does seem to be actual failure since last commit. Which is strange.

**Comment by brisvag** [comment]
Wow, that IS strange! I added a couple of prints just before the error:

```python
print(grid_widget)
print(type(grid_widget))
print(grid_widget['row'])
```
output:
```python
GridWidget(widget=<PyQt5.QtWidgets.QLabel object at 0x7ff080f21510>, row=0, column=0, row_span=1, column_span=1)
<class 'napari._qt.layer_controls.qt_layer_controls_base.GridWidget'>
Traceback (most recent call last):
    [... stuff ...]
    print(grid_widget['row'])
TypeError: tuple indices must be integers or slices, not str
```

What the hell?

**Comment by brisvag** [comment]
Well, it turns out I don't know how namedtuple works.

File: napari/_qt/layer_controls/qt_labels_controls.py
Comment (tlambert03) [inline napari/_qt/layer_controls/qt_labels_controls.py]: another quick question... this now looks a lot like a QFormLayout (label, widget).  Do you know whether that definitely wouldn't work here?

Comment (brisvag) [reply napari/_qt/layer_controls/qt_labels_controls.py]: Real answer: I didn't know it existed :P

Hindsight answer: Grid is a bit more versatile. The `button_row` from the code above, for example, takes 2 columns and 2 lines, which it seems wouldn't work with `QFormLayout`. We also can have things that are wider than 2 columns, or things that do not have a label (several examples of this in the code).

On the other hand, we could probably migrate to `QFormLayout` if we simplify some things, and maybe that's a good idea for the future :P

Comment (sofroniewn) [reply napari/_qt/layer_controls/qt_labels_controls.py]: Using a `QFormLayout` would be really nice. I think `button_row` might be the only place where this fails, and those button rows might be a bit funny anyway. If we could get everything in a `QFormLayout` and not need the elipses that would be huge.

One quick fix option might be to create a vertical layout, put the buttons at the top and then QFormLayout with everything else below it?

What do you think @tlambert03 @brisvag?

Comment (brisvag) [reply napari/_qt/layer_controls/qt_labels_controls.py]: If we're sure we won't need more complex behavour, that sounds loke a good plan to me.

Comment (sofroniewn) [reply napari/_qt/layer_controls/qt_labels_controls.py]: > If we're sure we won't need more complex behavour, that sounds loke a good plan to me.

i think constraints here are good

**Comment by brisvag** [comment]
I'm trying to convert this to `QFormLayout`. One thing I encountered in doing so is that we actually have the same issue as #3624, which was fixed in #3663 for plugins. I moved that piece of code deeper in the dock widget logic to solve the issue for everything.

However, I also noticed that the thing that prevents resizing the left side of the viewer is somehow tied to the layer controls. If you undock the layter controls, suddenly the handle appears and the left dock area can be resized! I could not for the life of me find what the controls have that's different from (say) the layerlist that may be causing this behavour. It must be caused by the fact that layercontrols is completely unresizeable, but I don't know what causes *that*....

**Comment by brisvag** [comment]
So, I migrated (just one layer for now) to the form layout. I found some issues here and there with resizing and fixed them. I tracked down a particularly annoying one that popped up in many other issues before:


https://user-images.githubusercontent.com/23482191/159722032-0a9a8b34-9ff8-4df5-982f-67d1ae0b17ae.mp4

As you can see, the `QSplitter` that is the main napari layout *should work*; the only thing that messes up with it is the layer controls widget. Unfortunately, despite my attempts, I could not find the place where the `MaximumSize` gets messed up. All the way to the end of the instantiation of `QtViewer`, that maximum size is set to very high numbers. Somewhere *after* that, things get messed up, but for some reason *only for layer controls*, and not for any of the other dock widgets. Any thoughts?

**Comment by sofroniewn** [comment]
> As you can see, the QSplitter that is the main napari layout should work; the only thing that messes up with it is the layer controls widget. Unfortunately, despite my attempts, I could not find the place where the MaximumSize gets messed up. All the way to the end of the instantiation of QtViewer, that maximum size is set to very high numbers. Somewhere after that, things get messed up, but for some reason only for layer controls, and not for any of the other dock widgets. Any thoughts?

I'm so sorry, maybe it's late, but I can't actually tell from the video what isn't working as expected. Can you describe really explicitly and then maybe I can help. When you move the vertical QSplitter left to right the controls expand left to right? Currently on main you can't actually move that QSplitter, no? What do you expect to happen?

**Comment by brisvag** [comment]
> When you move the vertical QSplitter left to right the controls expand left to right? Currently on main you can't actually move that QSplitter, no?

Exactly this! On main that `Qsplitter` is invisible (just lke at the beginning of the clip). With this PR, and additionally that line of code that I run in the console, the `QSplitter` appears and can be used. I expect the splitter to always be there.

**Comment by sofroniewn** [comment]
> I could not find the place where the MaximumSize gets messed up. All the way to the end of the instantiation of QtViewer, that maximum size is set to very high numbers. Somewhere after that, things get messed up, but for some reason only for layer controls, and not for any of the other dock widgets.

Is it here in the stylesheet?
 https://github.com/napari/napari/blob/f015c04df437834f95d7d3e8b0d81442bc76acac/napari/_qt/qt_resources/styles/02_custom.qss#L268-L269

**Comment by brisvag** [comment]
Oh, wow. I thought `css` was only about colors and stuff like that... This makes things a lot more complicated xD But at least it explains why so much was just kinda happening without a reason.

**Comment by brisvag** [comment]
I removed the upper limits to sizes of widgets; I don't see any reason why there should be an upper limit, but maybe @tlambert03 can chip in, since git `blame`s him? :)

Anyways, now I switched everything to forms, and the resizing works as expected. I'm happy to split things out in different PRs if people feel like these things should be separate.

**Comment by tlambert03** [comment]
Those limits were there before me, I just probably moved them from one file to another.  Fine with me to remove them

File: napari/_qt/menus/plugins_menu.py
Comment (sofroniewn) [inline napari/_qt/menus/plugins_menu.py:103]: Just curious why these changes are in this PR, are they needed?

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [inline napari/_qt/qt_main_window.py:873]: Ah ok you moved code here, that's probably fine - I'm just tracking

File: napari/_qt/layer_controls/qt_layer_controls_base.py
Comment (sofroniewn) [inline napari/_qt/layer_controls/qt_layer_controls_base.py]: can we remove commented line?

**Review by sofroniewn — APPROVED** [review]
Something funny happens when I launch napari with this PR - I now see the layer list widget appear independently for ~1 second before the main window opens - almost as if it is getting created ahead of time without a parent. That's probably the only major thing that needs fixing before merge

Generally though this is great - all the controls for all the layers still looked good to me

Before:

Screen Shot 2022-03-29 at 12 52 22 PM

After:
Screen Shot 2022-03-29 at 12 51 59 PM

You can see some fairly minor changes like the right align instead of the left align which I like. Would be nice for a couple other folks to try and break this too. Great work here @brisvag!

File: napari/_qt/qt_main_window.py
Comment (brisvag) [reply napari/_qt/qt_main_window.py:873]: Yup exactly, this change basically makes sure that that fix works for all dock widgets, not just plugin dock widgets. Still ugly, but it works.

File: napari/_qt/layer_controls/qt_layer_controls_base.py
Comment (brisvag) [reply napari/_qt/layer_controls/qt_layer_controls_base.py]: yup!

**Comment by brisvag** [comment]
Thanks @sofroniewn for checking. Not sure what's up with the layerlist widget, I don't see that happening on my side, but maybe it's just too fast. It might be caused by the code I moved with quickly toggles floating? Can you comment out those two lines and see if it still happens?

@jni, I feel like you usually have clear ideas about GUI and layouts; do you see anything off?

**Comment by brisvag** [comment]
> You can see some fairly minor changes like the right align instead of the left align which I like.

On my pc it's still left-aligned. I think it's because `QFormLayout` uses some OS-dependent defaults for this, and I have not set anything.

**Review by alisterburt — CHANGES_REQUESTED** [review]
took this for a quick spin with the volume plane rendering example (extra button row in 3d)
with this PR/main
imageimage

with this PR we have the extra layer control splitter and the layer controls are a bit messed up

Definitely not against the idea of simplifying the way layer controls are generated but this would need to be addressed before merge 🙂

**Comment by brisvag** [comment]
Adding the splitter back was intentional :)

As for the layout, It should be pretty easy to force the size of the widget to span the whole column, if that's what we want!

**Comment by brisvag** [comment]
> As for the layout, It should be pretty easy to force the size of the widget to span the whole column, if that's what we want!

Fixed :)

**Comment by brisvag** [comment]
Issues that are somwhat related to the resizing thing, and might be worth revisiting after the small changes here.

- https://github.com/napari/napari/issues/1710
- https://github.com/napari/napari/issues/3291
- https://github.com/napari/napari/issues/3620
- https://github.com/napari/napari/issues/2470
- https://github.com/napari/napari/issues/2337

**Comment by alisterburt** [comment]
Nice! Looking good for image controls in 3D now

I'm still seeing a lot of extra space between the layer controls and the layer list which wasn't there before
image

I'm also not convinced about the extra configurability of the layer controls and think I would prefer the autogeneration of the controls be a separate PR from the introduction of extra controls for resizing, what do you think about that?

**Comment by brisvag** [comment]
The extra space is just because the maximum height is no longer locked into place, and somewhere else there must be a "weaker" setting which defaults it to a different position. Not sure how to hunt for it...

What is it you don't like about being able to resize the controls? The fact that you couldn't so far was just a bug, and I don't see a reason why it shouldn't be allowed.

But yeah, I should definitely split these out in two PRs at this point :)

**Comment by brisvag** [comment]
Here's an issue that I couldn't solve: when rows in the `QFormLayout` are hidden, the widget spacing remains. This happens for example in the `Image` layer, where several widgets get hidden and shown between types. Here's a post with a few links about the issue, and how it's still basically unsolve: https://forum.qt.io/topic/29222/layout-spacing-when-widgets-are-hidden/4

The only solution I see is to actually remove the row, and readd it later; this however requires to hold the row position and juggle around with the indices depending on what gets added and what doesnt... Which kinda defeats the purpose of this PR.

So, are we ok with the little extra spacing between (for example) `rendering` and `depiction` in the `volume` layer in `examples/clipping_planes_interactive.py`? There' also probably a way around it by ordering them exactly right (I think the same issue would be on `main` as well); it's a just a bit annoying.

**Comment by brisvag** [comment]
I split out the resizing changes into #4368 and removed them from here.

File: napari/_qt/layer_controls/qt_image_controls.py
Comment (alisterburt) [inline napari/_qt/layer_controls/qt_image_controls.py:93]: Could you explain why piecing out these controls is necessary for this PR? The encapsulation of these controls in a single widget was specifically requested when they were added, maintainability being the cited reason
(No link, on mobile but PR was interactive plane controls)

**Review by alisterburt — COMMENTED** [review]
Thank you for piecing out the changes into separate PRs! 🙂

I've left a comment asking for clarification on why the plane controls needed piecing out, if we can't encapsulate more complex controls into their own widgets with this autogeneration that feels to me like an argument against autogeneration.

One other general feeling is that I'm not sure whether accepting visual inconsistencies for the sake of autogeneration (a 'nice to have' feature) is a good tradeoff. 

Overall I think these are good reasons not to merge here but don't feel too strongly about it and don't want to block if others feel like the benefits outweigh the negatives highlighted 🙂

**Comment by brisvag** [comment]
Just to clarify, the "autogeneration" was an earlier iteration of this PR. Now I'm just using `QFormLayout` which is a code `Qt` widget. So the code is not as obscure as it was before :P

> I've left a comment asking for clarification on why the plane controls needed piecing out, if we can't encapsulate more complex controls into their own widgets with this autogeneration that feels to me like an argument against autogeneration.

Yeah, this is the one thing that bothers me; maybe a solution is to (instead of having a single widget) putting all the "turn off/turn on" logic in a separate method?

> One other general feeling is that I'm not sure whether accepting visual inconsistencies for the sake of autogeneration (a 'nice to have' feature) is a good tradeoff.

I agree that we need to find a way to fix visual inconsistencies; maybe this can be fixed by overriding something deep in `Qt` so that the padding disappears if the widget is not visible.

However, note that here we're simply replacing one visual inconsistency with another. Here's main vs this pr: notice how the plane controls are out of line on main, which is fixed here. However, this PR adds the "spacing". By the way, we only have issues in the plane controls on the image layer.

image
image

> Overall I think these are good reasons not to merge here but don't feel too strongly about it and don't want to block if others feel like the benefits outweigh the negatives highlighted slightly_smiling_face

I'm ok with not merging if other people feel like this is not solving problems, so let's wait for other to chip in :).
I would like to point out that #4381 is attempting to add tabs to layer controls, and it might be usefult to think of that in this context. For example, it might be worth moving "more complex controls" (like the plane controls) to its own tab, which would solve some of our problems. On the other hand, tabs might make this completely obsolete :P

**Comment by alisterburt** [comment]
oh, weird! I had updated the qss so it looked right on my machine but your text looks larger in general - this is main on mac
image
https://github.com/napari/napari/blob/68d2c1b638c5481fde4364acb4abc4a8a4ce7c00/napari/_qt/qt_resources/styles/02_custom.qss#L295-L312

Not sure how to deal with getting this right on all platforms though :/

Will leave a comment on the tabbed controls at #4381

**Comment by jni** [comment]
> One other general feeling is that I'm not sure whether accepting visual inconsistencies for the sake of autogeneration (a 'nice to have' feature) is a good tradeoff.

Personally I think simple code is *very important* and from what I can see above, most of the visual issues have been addressed. What remains to be resolved here? The extra spacing at the bottom of the layer controls? I wonder if @sofroniewn or @tlambert03 would have ideas here? Basically we want the layer list to take as much space as is available vertically, right? I *think* that's a thing in Qt? 😂

**Comment by tlambert03** [comment]
> Basically we want the layer list to take as much space as is available vertically, right?

Do you mean that the height of both the controls and layerlist would be dynamic?  That actually could be pretty annoying;  imagine using the arrow keys to go through the layers, and the height of the layer-list is constantly changing.  Is that the current intention?

**Comment by jni** [comment]
@tlambert03 very good point! I forgot that each layer has different control heights. Hmm. Having said that, that means that we want to make the height of the controls the max of all the layer controls, which is a hard thing to determine dynamically... But maybe feasible? What do you think?

**Comment by tlambert03** [comment]
> Having said that, that means that we want to make the height of the controls the max of all the layer controls

yeah, I think that's kinda/sorta what we have now (just without closely considering what that height actually should minimally be), right?

> which is a hard thing to determine dynamically... But maybe feasible? What do you think?

Feasiblish ™️  :joy: 
I guess you'd have to determine which configuration of the controls was the tallest (e.g. "image controls with ndisplay = 3 and so on...") and then add up the vertical `sizeHints()` of all the visible widgets.
In practice, until these controls become _very_ dynamic in terms of the their contents, I suspect it's going to be easier to just manually pick a number and go with it (kinda how it currently is, but perhaps with a more conscious decision on the number)

**Comment by brisvag** [comment]
@jni @tlambert03 I agree that that could be useful, but that's actually unchanged in this PR (and therefore I would suggest to address it in another PR anyways). The issue about spacing that @alisterburt is referring to is actually different; see [this comment above](https://github.com/napari/napari/pull/4195#issuecomment-1091457618) for details :)

**Comment by brisvag** [comment]
In practice, we refer to this vertical gap here, above 'plane normal' (main first, this PR second):
image
image

**Comment by jni** [comment]
> In practice, we refer to this vertical gap here, above 'plane normal' (main first, this PR second):

Wow! Well, I really could not care less about it! 🤣 (Ok, technically I could, but imho this is *definitely* worth getting in, then. 😂)

**Comment by alisterburt** [comment]
> Just to clarify, the "autogeneration" was an earlier iteration of this PR. Now I'm just using QFormLayout which is a code Qt widget. So the code is not as obscure as it was before :P

sorry, I missed that the autogeneration had been removed, shouldn't review PRs on the move, my bad :) 

> What remains to be resolved here? The extra spacing at the bottom of the layer controls?

two small issues - both minor and non blocking
- the slight gap between the layer list and the plane controls
- the splitting out of plane controls widget into their individual components widgets

I only highlight the splitting out because this
> maybe a solution is to (instead of having a single widget) putting all the "turn off/turn on" logic in a separate method?

is how I originally implemented the plane controls and was asked to turn them into a single widget for maintainability/easy removal in future - I don't really mind either way :)

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
@alisterburt could you create a new review, maybe as a comment rather than requested changes if you really think it's non-blocking? 😉 That will give us the flexibility to include this in 0.4.16rc0.

**Review by alisterburt — APPROVED** [review]

**Comment by alisterburt** [comment]
done & done! 

edit: didn't realise my earlier comment review didn't 'unrequest' the changes

File: napari/_qt/layer_controls/qt_layer_controls_base.py
Comment (sofroniewn) [inline napari/_qt/layer_controls/qt_layer_controls_base.py]: We don't have to hold up the PR for this, but maybe just noting that this could now be `LayerFormLayout` instead and doc string could call out `form` instead of `grid`

File: napari/_qt/layer_controls/qt_layer_controls_base.py
Comment (sofroniewn) [inline napari/_qt/layer_controls/qt_layer_controls_base.py]: i wonder if we can just get the layout using `self.layout()` and then drop the `grid_layout` attribute (rather than rename to `form_layout`) cc @tlambert03 (note this is purely stylistic, but might as well fix now)

Comment (brisvag) [reply napari/_qt/layer_controls/qt_layer_controls_base.py]: I changed it to `layout()`; if anyone disagrees it's easy to revert :) Otherwise it's ready to merge!

**Comment by jni** [comment]
Boom! Thank you @brisvag and everyone who participated in the various review iterations! :tada:

**Pull request merged**
